src/context.cpp

```diff
@@ -454,24 +455,48 @@ namespace librealsense
         auto callback_id = unique_id::generate_id();
         _devices_changed_callbacks.insert(std::make_pair(callback_id, std::move(callback)));
 
+        if (_device_watcher_active == false)
+        {
+            _device_watcher->stop();
+
+            _device_watcher->start([this](platform::backend_device_group old, platform::backend_device_group curr)
+            {
+                on_device_changed(old, curr, _playback_devices, _playback_devices);
+            });
+            _device_watcher_active = true;
+        }
+
         return callback_id;
     }
 
     void context::unregister_internal_device_callback(uint64_t cb_id)
     {
         std::lock_guard<std::mutex> lock(_devices_changed_callbacks_mtx);
         _devices_changed_callbacks.erase(cb_id);
+
+        if (_device_watcher_active && _devices_changed_callback == nullptr && _devices_changed_callbacks.size() == 0)
+        {
+            _device_watcher->stop();
+            _device_watcher_active = false;
+        }
     }
 
     void context::set_devices_changed_callback(devices_changed_callback_ptr callback)
     {
-        _device_watcher->stop();
-
+        std::lock_guard<std::mutex> lock(_devices_changed_callbacks_mtx);
         _devices_changed_callback = std::move(callback);
-        _device_watcher->start([this](platform::backend_device_group old, platform::backend_device_group curr)
+
+        if (_device_watcher_active == false)
```

Comment:
This should all go in a private function `start_device_watcher()`

Reply:
done